### ISSUE TITLE
Provide `tester` parameter to `setUp` callbacks

### ISSUE THREAD

**Issue by spkersten** [flutter/flutter#52003]
I would like to separate tests (expectations/verifications) from the test setup. A common structure for tests is:

```dart
group("when X is done", () {
  setUp(() {
    foo.X();
  });

  test("then Y must be done", (){
    verify(Y()).called(1);
  });

  group("when A is done next", () {
    setUp(() {
       foo.A();
    });
    
    test("then B must be done twice", () {
      verify(B()).called(2);
    });
  });
});
```

With widget tests this is not possible, because no `WidgetTester` object is passed to `setUp` methods to interact with. So, I can't pump a widget in a setUp method, not perform some interactions to get the widget into the proper initial state.

**Comment by spkersten** [comment]
I've created this wrapper as a somewhat crude way of offering this functionality: https://pub.dev/packages/flutter_test_ui

**Comment by mehmetf** [comment]
@spkersten I am sorry to say that your use case won't be covered by https://github.com/flutter/flutter/pull/57894. I agree with @hixie that we should actively try to reduce the fragmentation between package:test and package:flutter_test. That means *not* introducing additional setup functions i.e. setup/teardown functions won't be accepting a WidgetTester argument.

I understand there are different approaches to organizing testing, thus your flutter_test_ui package is a welcome addition.

Within core Flutter team, we would recommend the pattern of using helper methods to setup widgets directly from your testWidgets fn. We believe test() methods should be self-describing. I should not need to look at two (or more) different places to get a complete picture of the test.

So what is setup used for then?

Typically setup functions are used to prep infra that individual tests (and their authors) should not need to care. For instance, I have some widget tree I want to test but every widget in my app directly or indirectly needs some `AuthService` to be present. The key point here is that **my test isn't about `AuthService`**. It just happens to be a dependency of my test. I can choose to set that up via `setUp` function to hide it from `test` and make it easier for test authors to add tests because they don't need to care about that dependency.

As such, I will be closing this since we won't be fixing it.

**Comment by spkersten** [comment]
@mehmetf 

> The key point here is that my test isn't about AuthService. It just happens to be a dependency of my test. I can choose to set that up via setUp function to hide it from test and make it easier for test authors to add tests because they don't need to care about that dependency.

Similarly, the key point of many tests isn't certain activities that need `tester` (e.g. tapping a widget to expand it, so a test can do something with its content).

> Within core Flutter team, we would recommend the pattern of using helper methods to setup widgets directly from your testWidgets fn. We believe test() methods should be self-describing. I should not need to look at two (or more) different places to get a complete picture of the test.

It's frustrating to see this (and many other) issues so easily dismissed. Developers working as a team on a large app have different needs than those working on the flutter framework. 

Maybe on a higher level: From the related PR, I get that there are mainly concerns with the amount of effort and lack of backwards compatibility and not so much with the usefulness of the feature. It would be less unsatisfying if that was acknowledged as the reason for rejecting the feature request instead of trying to argue why the feature isn't useful anyway (while it clear is in our code base).

**Comment by mehmetf** [comment]
> Similarly, the key point of many tests isn't certain activities that need tester (e.g. tapping a widget to expand it, so a test can do something with its content).

That is a slippery slope. That pattern encourages tests that are simply asserts rather than also describing the flow they are testing. This leads to fragile tests where the logic of the test is distributed across several code blocks that are not immediately apparent. 

Note that what I am saying applies to all `setUp` function usage. This is one reason why Hixie's opinion is that we should just remove it.

> Developers working as a team on a large app have different needs than those working on the flutter framework.

I completely agree. I am **not** part of the core Flutter team. I support Flutter Google apps. We have a few very large apps (~1M LOC) and many medium sized ones (~50-200kLOC). I don't take these decisions lightly. We measure the usefulness of a feature by talking to leads of these apps and looking at how many votes an issue has gotten on github. 

The question here is whether this feature should be supported with a package (like yours) or should be part of the framework.  There are many features that are better suited for a package. Flutter itself was designed to be endlessly extensible. 

In some cases, features are part of the framework. For instance, I am wholly against removing setup/teardown entirely. It is a well-loved feature and will create endless questions/confusion about why a very popular testing pattern is not available in Flutter. 

In this case, we have made the decision that the cost of this feature (API fragmentation, ongoing maintenance, confusion over existing setup/teardown functions, encouraging obscure widget test setup) is not worth the benefit it provides. 

However, that's just the framework recommendation. The community is welcome to build a package that fits their needs, as you have done.

PS: I have reviewed your flutter_test_ui package and I don't think it is hacky. You had to replicate the test() API which is relatively stable and can be augmented based on user needs by adding more optional named parameters. I think it is a fine package.

PPS: The amount of effort to implement did not factor into this decision. The original bug I am trying to solve (setup/teardown execute in different Zones than test) is actually much harder as it will require changes to core test package.
